src/rectangle.cpp

```diff
@@ -79,63 +41,64 @@ static RectList* insert_rect_border(RectList* head,
     // |------+--------+-------|
     // |        bottom         |
     // +-----------------------+
-    RectList *top, *left, *right, *bottom;
     // coordinates of the bottom right corner of large
     int br_x = large.x + large.width, br_y = large.y + large.height;
-    RectList* (*r)(int,int,int,int) = rectlist_create_simple;
-    top   = r(large.x, large.y, large.x + large.width, center.y);
-    left  = r(large.x, center.y, center.x, center.y + center.height);
-    right = r(center.x + center.width, center.y, br_x, center.y + center.height);
-    bottom= r(large.x, center.y + center.height, br_x, br_y);
-
-    RectList* parts[] = { top, left, right, bottom };
-    for (unsigned int i = 0; i < LENGTH(parts); i++) {
-        head = reclist_insert_disjoint(head, parts[i]);
-    }
-    return head;
-}
-
-// insert a new element without any intersections into the given list
-RectList* reclist_insert_disjoint(RectList* head, RectList* element) {
-    if (!element) {
-        return head;
-    } else if (!head) {
-        // if the list is empty, then intersection-free insertion is trivial
-        element->next = nullptr;
-        return element;
-    } else if (!rects_intersect(head->rect, element->rect)) {
-        head->next = reclist_insert_disjoint(head->next, element);
-        return head;
-    } else {
-        // element intersects with the head rect
-        auto center = intersection_area(head->rect, element->rect);
-        auto large = head->rect;
-        head->rect = center;
-        head->next = insert_rect_border(head->next, large, center);
-        head->next = insert_rect_border(head->next, element->rect, center);
-        g_free(element);
-        return head;
+    Rectangle (*r)(int,int,int,int) = Rectangle::fromCorners;
```

Comment:
Unless you prefer it like it this, I *think* you can just say `auto r = ...` here.